examples/herokilled/src/handler.rs

```diff
@@ -0,0 +1,157 @@
+    }
+}
+
+trait Handler<S> {
+    fn handle(
+        &mut self,
+        state: &mut S,
+        ctx: &Context,
+        packet_type: u32,
+        data: &[u8],
+    ) -> parser::Result<()>;
+}
+
+struct HandlerImpl<S: PlayerState, M: prost::Message + Default, H: MessageHandler<S, M>> {
+    id: u32,
+    handler: Box<H>,
+    _phantom1: PhantomData<S>,
+    _phantom2: PhantomData<M>,
+}
+
+impl<S: PlayerState, M: prost::Message + Default, H: MessageHandler<S, M>> HandlerImpl<S, M, H> {
+    pub fn new(id: u32, handler: H) -> Box<Self> {
+        Box::new(Self {
+            id,
+            handler: Box::new(handler),
+            _phantom1: PhantomData,
+            _phantom2: PhantomData,
+        })
+    }
+}
+
+impl<S: PlayerState, M: prost::Message + Default, H: MessageHandler<S, M>> Handler<S>
+    for HandlerImpl<S, M, H>
+{
+    fn handle(
+        &mut self,
+        state: &mut S,
+        ctx: &Context,
+        packet_type: u32,
+        data: &[u8],
+    ) -> parser::Result<()> {
+        if packet_type != self.id {
+            return Ok(());
+        }
+
+        let msg = M::decode(data)?;
+
+        self.handler.handle(state, ctx, &msg)
+    }
+}
+
+pub struct Player {
+    pub slot: i32,
+    pub team_id: u8,
+    pub name: String,
+}
+
+pub struct HandlerVisitor<S: PlayerState> {
+    state: S,
+    handlers: Vec<Box<dyn Handler<S>>>,
+}
+
+impl<S: PlayerState + 'static> HandlerVisitor<S> {
+    pub fn with_state(state: S) -> Self {
+        Self {
+            state,
+            handlers: vec![],
+        }
+    }
+
+    pub fn with<M: prost::Message + Default + 'static, H: MessageHandler<S, M> + 'static>(
+        mut self,
+        id: u32,
+        handler: H,
+    ) -> Self {
+        self.handlers.push(HandlerImpl::new(id, handler));
+
+        self
+    }
+
+    pub fn state(&self) -> &S {
+        &self.state
+    }
+}
+
+impl<S: PlayerState> Visitor for &mut HandlerVisitor<S> {
+    fn on_packet(&mut self, ctx: &Context, packet_type: u32, data: &[u8]) -> parser::Result<()> {
+        for handler in self.handlers.iter_mut() {
+            handler.handle(&mut self.state, ctx, packet_type, data)?
+        }
+
+        Ok(())
+    }
+
+    fn on_entity(
+        &mut self,
+        ctx: &Context,
+        update_flags: usize,
+        update_type: crate::entities::UpdateType,
+        // TODO: include updated fields (list of field paths?)
+        entity: &crate::entities::Entity,
+    ) -> parser::Result<()> {
+        let team_key: u64 = entities::make_field_key(&["m_iTeamNum"]);
+        let teamname_key: u64 = entities::make_field_key(&["m_szTeamname"]);
+        let playername_key: u64 = entities::make_field_key(&["m_iszPlayerName"]);
+        let playerslot_key: u64 = entities::make_field_key(&["m_unLobbyPlayerSlot"]);
+
+        let ser = entity.get_serializer().serializer_name.str.to_string();
+
+        if ser != "CCitadelPlayerController" {
+            return Ok(());
+        }
```

Comment:
it would be more optimal to hash `CCitadelPlayerController` using `fxhash::hash_bytes` and compare hashes instead of strings.

`preserve-metadata` feature is intended to be useful only for debugging or building tools for demos.

Reply:
tangential question - I had the field names hashes as consts and got a compilation error. will recreate it to find the exact error, but it definitely wasn't happy

Reply:
```
error[E0080]: evaluation of constant value failed
   --> /home/pwf/src/haste/crates/haste/src/fxhash.rs:80:38
    |
80  |         hash = add_u64_to_hash(hash, blocks[i >> 3]);
    |                                      ^^^^^^^^^^^^^^ accessing memory based on pointer with alignment 1, but alignment 8 is required
```

Reply:
would this work for compile time hashing? https://github.com/HindrikStegenga/const-fnv1a-hash/blob/main/src/lib.rs